narwhal/primary/src/state_handler.rs

```diff
@@ -185,16 +255,195 @@ impl RandomnessState {
+                        error!("random beacon: unable to write VSS key to output: {e:?}")
+                    }
+                    let num_shares = output.shares.as_ref().map_or(0, |shares| shares.len());
                     self.dkg_output = Some(output);
+                    info!("random beacon: DKG complete with {num_shares} shares for this node");
+                }
+                Err(fastcrypto::error::FastCryptoError::NotEnoughInputs) => (), // wait for more input
+                Err(e) => error!("random beacon: error while processing DKG Confirmations: {e:?}"),
+            }
+            // Begin randomness generation.
+            if self.dkg_output.is_some() {
+                info!("random beacon: start randomness generation");
+                self.send_partial_signatures().await;
+            }
+        }
+    }
+
+    async fn update_narwhal_round(&mut self, round: Round) {
+        self.narhwal_round = round;
+        let next_leader_narwhal_round = round + (round % 2);
+        // Re-send partial signatures to new leader, in case the last one failed.
+        if self.dkg_output.is_some() && (next_leader_narwhal_round > self.last_narwhal_round_sent) {
+            self.send_partial_signatures().await;
+        }
+    }
+
+    async fn update_randomness_round(&mut self, round: RandomnessRound) {
+        if round <= self.randomness_round {
+            // Don't go backwards.
+            return;
+        }
+        self.randomness_round = round;
+        self.partial_sigs
+            .retain(|&(round, _), _| round >= self.randomness_round);
+        self.send_partial_signatures().await;
+    }
+
+    async fn send_partial_signatures(&mut self) {
+        let dkg_output = match &self.dkg_output {
+            Some(dkg_output) => dkg_output,
+            None => {
+                error!("random beacon: called send_partial_signatures before DKG completed");
+                return;
+            }
+        };
+        let shares = match &dkg_output.shares {
+            Some(shares) => shares,
+            None => return, // can't participate in randomness generation without shares
+        };
+        debug!(
+            "random beacon: sending partial signatures for round {}",
+            self.randomness_round
+        );
+        if self.cached_sigs.is_none()
+            || self.cached_sigs.as_ref().unwrap().0 != self.randomness_round
+        {
+            self.cached_sigs = Some((
+                self.randomness_round,
+                ThresholdBls12381MinSig::partial_sign_batch(
+                    shares,
+                    &self.randomness_round.signature_message(),
+                ),
+            ));
+        }
+        let sigs = self.cached_sigs.as_ref().unwrap().1.clone();
+
+        let next_leader_narwhal_round = self.narhwal_round + (self.narhwal_round % 2);
+        self.last_narwhal_round_sent = next_leader_narwhal_round;
+
+        let leader = self.leader_schedule.leader(next_leader_narwhal_round);
```

Comment:
When self.narhwal_round is even, next_leader_narwhal_round will be computed to be the same value. Sending a message to the leader at this round will be a bit too late, because the leader message is likely to have been broadcasted. Maybe the calculation should have been the +2 round leader, or the destination should be deterministic on next randomness_round instead?